keps/sig-api-machinery/20190206-watch-bookmark.md

```diff
@@ -0,0 +1,184 @@
+---
+title: Watch Bookmark
+authors:
+  - "@wojtek-t"
+owning-sig: sig-apimachinery
+participating-sigs:
+  - sig-scalability
+reviewers:
+  - "@jpbetz"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-02-06
+status: implementable
+see-also:
+  - https://github.com/kubernetes/kubernetes/issues/73585
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Watch bookmark
+
+## Table of Contents
+
+Table of Contents
+=================
+
+* [Watch bookmark](#watch-bookmark)
+   * [Table of Contents](#table-of-contents)
+   * [Summary](#summary)
+   * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-Goals](#non-goals)
+   * [Proposal](#proposal)
+      * [Risks and Mitigations](#risks-and-mitigations)
+   * [Graduation Criteria](#graduation-criteria)
+   * [Implementation History](#implementation-history)
+   * [Rejected alternatives](#rejected-alternatives)
+      * [Cache in kube-apiserver](#cache-in-kube-apiserver)
+      * [API for send bookmark](#api-for-send-bookmark)
+
+## Summary
+
+Watch API is one of the fundaments of Kubernetes API. The recommended pattern
+for using watch API is to retrieve a collection of resources using consistent
+list and then initiate a watch starting from a resourceVersion returned by the
+list operation. If the client watch is disconnected, a new one can be restarted
+from the last returned resourceVersion.
+
+This proposal make restarting watches cheaper from kube-apiserver performance
+perspective.
+
+## Motivation
+
+While running different scalability tests we observed that restarting watches
+may cause significant load on kube-apiserver when watcher is observing a small
+percentage of changes (due to field or label selector). In extreme cases,
```

Comment:
Does it refer to a similar case mentioned below？
> Kubelet for watching pods for a release

Bookmark is reasonable optimization for watch. In the kubelet case, list pods by node-name is very generic, we can consider adding an index to optimize the list operation. If that make sense to you，I can create a new issue for it.

Reply:
Watchcache already has an index for it :-)